File: eval-server/python/src/bo_eval_server/evaluation_stack.py
"""
EvaluationStack - A simple stack-like structure for managing evaluations.

Provides LIFO (Last In, First Out) access to evaluation objects.
Useful for distributing different evaluations across multiple client connections.
"""

from typing import Dict, Any, List, Optional


class EvaluationStack:
    """A LIFO stack for managing evaluation objects."""
    
    def __init__(self) -> None:
        """Initialize an empty evaluation stack."""
        self._evaluations: List[Dict[str, Any]] = []
    
    def push(self, evaluation: Dict[str, Any]) -> None:
        """
        Add an evaluation to the top of the stack.
        
        Args:
            evaluation: The evaluation object to add
            
        Raises:
            ValueError: If evaluation is invalid or missing required fields
        """
        if not evaluation or not isinstance(evaluation, dict):
            raise ValueError('Evaluation must be a valid dictionary')
        
        # Validate required fields
        required_fields = ['id', 'name', 'tool', 'input']
        for field in required_fields:
            if field not in evaluation or not evaluation[field]:
                raise ValueError(f'Evaluation missing required field: {field}')
        
        self._evaluations.append(evaluation)
    
    def pop(self) -> Optional[Dict[str, Any]]:
        """
        Remove and return the evaluation from the top of the stack.
        
        Returns:
            The evaluation object, or None if stack is empty
        """
        if self._evaluations:
            return self._evaluations.pop()
        return None
    
    def is_empty(self) -> bool:
        """
        Check if the stack is empty.
        
        Returns:
            True if stack has no evaluations
        """
        return len(self._evaluations) == 0
    
    def size(self) -> int:
        """
        Get the number of evaluations in the stack.
        
        Returns:
            The stack size
        """
        return len(self._evaluations)
    
    def peek(self) -> Optional[Dict[str, Any]]:
        """
        Peek at the top evaluation without removing it.
        
        Returns:
            The top evaluation object, or None if stack is empty
        """
        if self.is_empty():
            return None
        return self._evaluations[-1]
    
    def clear(self) -> None:
        """Clear all evaluations from the stack."""
        self._evaluations.clear()
    
    def to_array(self) -> List[Dict[str, Any]]:
        """
        Get a copy of all evaluations in the stack (top to bottom).
        
        Returns:
            List of evaluation objects from top to bottom
        """
        return list(reversed(self._evaluations))
    
    def __len__(self) -> int:
        """Return the number of evaluations in the stack."""
        return len(self._evaluations)
    
    def __bool__(self) -> bool:
        """Return True if stack has evaluations."""
        return not self.is_empty()
    
    def __repr__(self) -> str:
        """String representation of the stack."""
        return f"EvaluationStack(size={self.size()})"

File: eval-server/nodejs/src/lib/EvaluationStack.js
// Copyright 2025 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

/**
 * EvaluationStack - A simple stack-like structure for managing evaluations
 * 
 * Provides LIFO (Last In, First Out) access to evaluation objects.
 * Useful for distributing different evaluations across multiple client connections.
 */
export class EvaluationStack {
  constructor() {
    this.evaluations = [];
  }

  /**
   * Add an evaluation to the top of the stack
   * @param {Object} evaluation - The evaluation object to add
   */
  push(evaluation) {
    if (!evaluation || typeof evaluation !== 'object') {
      throw new Error('Evaluation must be a valid object');
    }
    
    // Validate required fields
    const requiredFields = ['id', 'name', 'tool', 'input'];
    for (const field of requiredFields) {
      if (!evaluation[field]) {
        throw new Error(`Evaluation missing required field: ${field}`);
      }
    }
    
    this.evaluations.push(evaluation);
  }

  /**
   * Remove and return the evaluation from the top of the stack
   * @returns {Object|null} The evaluation object, or null if stack is empty
   */
  pop() {
    return this.evaluations.pop() || null;
  }

  /**
   * Check if the stack is empty
   * @returns {boolean} True if stack has no evaluations
   */
  isEmpty() {
    return this.evaluations.length === 0;
  }

  /**
   * Get the number of evaluations in the stack
   * @returns {number} The stack size
   */
  size() {
    return this.evaluations.length;
  }

  /**
   * Peek at the top evaluation without removing it
   * @returns {Object|null} The top evaluation object, or null if stack is empty
   */
  peek() {
    if (this.isEmpty()) {
      return null;
    }
    return this.evaluations[this.evaluations.length - 1];
  }

  /**
   * Clear all evaluations from the stack
   */
  clear() {
    this.evaluations = [];
  }

  /**
   * Get a copy of all evaluations in the stack (top to bottom)
   * @returns {Array} Array of evaluation objects
   */
  toArray() {
    return [...this.evaluations].reverse();
  }
}